File: SenLib/Sen3/FileFixes/t_magic.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	internal class MagicData {
		public byte[] d0;
		public string flags;
		public byte[] d1;
		public string animation;
		public string name;
		public string desc;
		public byte[] d2;

		internal MagicData(byte[] data) {
			var stream = new DuplicatableByteArrayStream(data);
			d0 = stream.ReadBytes(4);
			flags = stream.ReadUTF8Nullterm();
			d1 = stream.ReadBytes(0x66);
			animation = stream.ReadUTF8Nullterm();
			name = stream.ReadUTF8Nullterm();
			desc = stream.ReadUTF8Nullterm();
			d2 = stream.ReadBytes(stream.Length - stream.Position);
		}

		internal byte[] ToBinary() {
			MemoryStream ms = new MemoryStream();
			ms.Write(d0);
			ms.WriteUTF8Nullterm(flags);
			ms.Write(d1);
			ms.WriteUTF8Nullterm(animation);
			ms.WriteUTF8Nullterm(name);
			ms.WriteUTF8Nullterm(desc);
			ms.Write(d2);
			return ms.CopyToByteArrayAndDispose();
		}
	}

	class t_magic : FileMod {
		public string GetDescription() {
			return "Fix formatting issues in art/craft descriptions.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xc0b07b04d56268a7ul, 0xc42471d6671dc5cbul, 0x6243286fu));
			if (file_en == null) {
				return null;
			}
			var tbl_en = new Tbl(file_en, EndianUtils.Endianness.LittleEndian);

			//for (int i = 0; i < tbl_en.Entries.Count; ++i) {
			//	var e = tbl_en.Entries[i];
			//	if (e.Name == "magic") {
			//		var m = new MagicData(e.Data);
			//		m.flags = m.flags.Replace("Z", "");
			//		m.name = m.name.Replace("\n", " ");
			//		e.Data = m.ToBinary();
			//	}
			//}

			// a bunch of missing or extra spaces
			{
				var m = new MagicData(tbl_en.Entries[64].Data);
				m.desc = m.desc.InsertSubstring(62, " ", 0, 1);
				tbl_en.Entries[64].Data = m.ToBinary();
			}
			{
				var m = new MagicData(tbl_en.Entries[126].Data);
				m.desc = m.desc.Remove(83, 1);
				tbl_en.Entries[126].Data = m.ToBinary();
			}
			for (int i = 0; i < tbl_en.Entries.Count; ++i) {
				var e = tbl_en.Entries[i];
				if (e.Name == "magic") {
					var m = new MagicData(e.Data);
					if (m.desc.Contains(" -#")) {
						m.desc = m.desc.Replace(" -#", " - #");
						e.Data = m.ToBinary();
					}
					if (m.desc.Contains(" #0C ")) {
						m.desc = m.desc.Replace(" #0C ", "#0C ");
						e.Data = m.ToBinary();
					}
				}
			}

			// normalize newlines
			for (int i = 0; i < tbl_en.Entries.Count; ++i) {
				var e = tbl_en.Entries[i];
				if (e.Name == "magic") {
					var m = new MagicData(e.Data);
					if (m.desc.StartsWith("[") && m.flags.Contains("Z") && m.desc.Contains("\n")) {
						m.desc = t_item.AdjustNewlinesToTwoSpaces(m.desc);
						e.Data = m.ToBinary();
					}
				}
			}

			Stream result_en = new MemoryStream();
			tbl_en.WriteToStream(result_en, EndianUtils.Endianness.LittleEndian);
			DuplicatableByteArrayStream result_en_b = result_en.CopyToByteArrayStream();
			return new FileModResult[] {
				new FileModResult("data/text/dat_en/t_magic.tbl", result_en_b.Duplicate())
			};
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file_en = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xc0b07b04d56268a7ul, 0xc42471d6671dc5cbul, 0x6243286fu));
			if (file_en == null) {
				return null;
			}
			return new FileModResult[] {
				new FileModResult("data/text/dat_en/t_magic.tbl", file_en)
			};
		}
	}
}

File: native/sen3/FileFixes/t_magic.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen3/file_fixes.h"
#include "sen3/tbl.h"
#include "sha1.h"
#include "util/text.h"

namespace SenLib::Sen3::FileFixes::t_magic {
std::string_view GetDescription() {
    return "Fix formatting issues in art/craft descriptions.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    using namespace HyoutaUtils::TextUtils;

    try {
        auto file = getCheckedFile(
            "data/text/dat_en/t_magic.tbl",
            85506,
            SenPatcher::SHA1FromHexString("c0b07b04d56268a7c42471d6671dc5cb6243286f"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;
        Tbl tbl_en(bin.data(), bin.size(), HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // for (size_t i = 0; i < tbl_en.Entries.size(); ++i) {
        //     auto& e = tbl_en.Entries[i];
        //     if (e.Name == "magic") {
        //         MagicData m(e.Data.data(), e.Data.size());
        //         m.flags = Replace(m.flags, "Z", "");
        //         m.name = Replace(m.name, "\n", " ");
        //         e.Data = m.ToBinary();
        //     }
        // }

        // a bunch of missing or extra spaces
        {
            auto& e = tbl_en.Entries[64];
            MagicData m(e.Data.data(), e.Data.size());
            m.desc = InsertSubstring(m.desc, 64, " ", 0, 1);
            e.Data = m.ToBinary();
        }
        {
            auto& e = tbl_en.Entries[126];
            MagicData m(e.Data.data(), e.Data.size());
            m.desc = Remove(m.desc, 83, 1);
            e.Data = m.ToBinary();
        }
        for (size_t i = 0; i < tbl_en.Entries.size(); ++i) {
            auto& e = tbl_en.Entries[i];
            if (e.Name == "magic") {
                MagicData m(e.Data.data(), e.Data.size());
                if (m.desc.find(" -#") != std::string::npos) {
                    m.desc = Replace(m.desc, " -#", " - #");
                    e.Data = m.ToBinary();
                }
                if (m.desc.find(" #0C ") != std::string::npos) {
                    m.desc = Replace(m.desc, " #0C ", "#0C ");
                    e.Data = m.ToBinary();
                }
            }
        }

        // Radiant Wings seems to have its description incorrectly copied from CS2
        // PS4 patch 1.03 changes it to "Stokes the fighting spirit, increasing one's strength.",
        // use that
        {
            const auto get_name_substr =
                [&](size_t idx, size_t offset, size_t length) -> std::string {
                auto& e = tbl_en.Entries[idx];
                return MagicData(e.Data.data(), e.Data.size()).name.substr(offset, length);
            };
            const auto get_desc_substr =
                [&](size_t idx, size_t offset, size_t length) -> std::string {
                auto& e = tbl_en.Entries[idx];
                return MagicData(e.Data.data(), e.Data.size()).desc.substr(offset, length);
            };

            auto& e = tbl_en.Entries[89];
            MagicData m(e.Data.data(), e.Data.size());
            std::string s = "";
            s += get_name_substr(109, 5, 3);
            s += get_desc_substr(152, 147, 8);
            s += get_desc_substr(35, 86, 15);
            s += get_desc_substr(24, 115, 4);
            s += m.desc.substr(87, 5);
            s += get_desc_substr(168, 37, 6);
            s += get_desc_substr(231, 30, 4);
            s += m.desc.substr(95, 8);
            s += m.desc.back();
            m.desc = ReplaceSubstring(m.desc, 85, 56, s, 0, s.size());
            e.Data = m.ToBinary();
        }

        // normalize newlines
        for (size_t i = 0; i < tbl_en.Entries.size(); ++i) {
            auto& e = tbl_en.Entries[i];
            if (e.Name == "magic") {
                MagicData m(e.Data.data(), e.Data.size());
                if (m.desc.starts_with("[") && m.flags.find_first_of('Z') != std::string::npos
                    && m.desc.find_first_of('\n') != std::string::npos) {
                    m.desc = AdjustNewlinesToTwoSpaces(std::move(m.desc));
                    e.Data = m.ToBinary();
                }
            }
        }

        std::vector<char> result_en_vec;
        MemoryStream result_en(result_en_vec);
        tbl_en.WriteToStream(result_en, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(result_en_vec), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::t_magic
